File: Data Structure Basics/Stacks/Java Implementation/Java Implementation/src/StackImplementation.java
import java.util.*;
import java.util.stream.Collectors;

public class StackImplementation
{
    public static List<Integer> execute(List<String> program)
    {
        // initialize stack
        ArrayList<Integer> stack = new ArrayList<>();

        for (String instruction: program)
        {
            if (instruction.equals("peek"))
            {
                // print out the top item in stack
                System.out.println(stack.get(stack.size()-1));
            }
            else if (instruction.equals("pop"))
            {
                //pop the top item in stack
                stack.remove(stack.size() - 1 );
            }
            else
            {
              // get the data in the "push" instruction
              int data = Integer.parseInt(instruction.substring(5));
              // push data to the top of stack
                stack.add(data);
            }
        }
        return stack;
    }

    public static void main(String[] args)
    {
        Scanner scanner = new Scanner(System.in);
        int programLength = Integer.parseInt(scanner.nextLine());
        List<String> program = new ArrayList<>();
        for (int i = 0; i < programLength; i++)
        {
            program.add(scanner.nextLine());
        }
        scanner.close();
        List<Integer> res = execute(program);
        System.out.println(res.stream().map(String::valueOf).collect(Collectors.joining(" ")));
    }
}

File: Data Structure Basics/Stacks/JS Implementation/StackImplementation.js
function execute(program)
{
    //initialize stack
    const stack = [];
    for (const instruction  of program)
    {
        if (instruction == "peek")
        {
            //print out top item in stack
            console.log(stack[stack.length-1]);
        }
        else if (instruction == "pop")
        {
            //pop the top item in stack
            stack.pop();
        }
        else
        {
            // get the data in the "push" instruction
            const data = parseInt(instruction.substring(5));
            // push data to the top of stack
            stack.push(data);
        }
    }
    return stack;
}

function* main()
{
    const programLength = parseInt(yield );
    const program = [];
    for (let i = 0; i < programLength; i++)
    {
        program.push(yield);
    }
    const res = execute(program);
    console.log(res.join(' '));
}

class EOFError extends Error {}
{
    const gen = main();
    const next = (line) => gen.next(line).done && process.exit();
    let buf = " ";
    next();
    process.stdin.setEncoding('utf8');
    process.stdin.on('data',(data) =>
    {
        const lines = (buf + data).split('\n');
        buf = lines.pop();
        lines.forEach(next);
    });
    process.stdin.on('end', () =>
    {
        buff && next(buf);
        gen.throw(new EOFError());
    })

}